File: Java/Vigenere.java

public class Vigenere extends Frase implements X{
	private String Key; //Chiave
    private char[] Risultato = new char[200]; //Testo risultante
    private char[] CipherString = new char[200]; //Stringa di cifratura
    private int[] S = new int[200]; //Vettore contenente il numero di spostamenti per ogni lettera
    
    /* Controllo se la chiave inserita rispetta i vincoli imposti (ovvero se e' costituita da soli caratteri alfabetici) e in
     * caso affermativo trasformo ogni lettera da minuscolo a maiuscolo */
	public Vigenere(String phrase, String key) throws Errori{
		super(phrase);
		String temp = key.toUpperCase();
		for(int i=0;i<key.length();i++) {
			if(temp.charAt(i)<'A' || temp.charAt(i)>'Z') {
				this.Key = "";
				throw new Errori("La chiave deve contenere solo caratteri alfabetici senza spazi!");
			}
		}
		this.Key = temp;
	}
	/* Ritorno la stringa di cifratura */
	public char[] getCipherString() {
		return CipherString;
	}
	/* Ritorno il testo risultante */
	public char[] getRisultato() {
		return Risultato;
	}
	/* Cifro il testo in chiaro */
	public void codifica() {
        buildStringVig();
        String f = getPhrase();
        for(int i=0;i<f.length();i++){
            if(S[i] != 0) Risultato[i] = getLettera(f.charAt(i),S[i]);
            else Risultato[i] = f.charAt(i);
        }
    }
	/* Decifro il testo cifrato */
	public void codificaRev() {
        buildStringVig();
        String f = getPhrase();
        for(int i=0;i<f.length();i++){
            if(S[i] != 0) Risultato[i] = getLetteraRev(f.charAt(i),S[i]);
            else Risultato[i] = f.charAt(i);
        }
    }
	/* Costruisco la stringa di cifratura e l'array che conterra' il numero di spostamenti da effettuare per ciascuna lettera
	   del testo inserito */
	public void buildStringVig(){
        String f = getPhrase();
        char[] Lett = getAlfa();
        for(int i=0,j=0;i<f.length();i++){
            if(f.charAt(i)!=' '){
                if(j>=Key.length()) j=0;
                CipherString[i] = Key.charAt(j);
                j++;
            }
            else CipherString[i] = ' ';
        }
        for(int i=0;i<f.length();i++){
            if(CipherString[i] != ' '){
                for(int j=0;j<26;j++){
                    if(CipherString[i] == Lett[j]) S[i] = j;
                }
            }
            else S[i] = 0;
        }
    }
	/* Trovo e ritorno la lettera che si trova 'pos' posizioni prima rispetto a 'c' */
	public char getLetteraRev(char c,int pos) {
        char l;
        char[] alfa = getAlfa();
        int temp = 0;
        for(int i=0;i<26;i++){
            if(alfa[i]==c){
                temp = i;
                break;
            }
        }
        if(temp >= pos)
            l = alfa[temp-pos];
        else{
            pos = 26 + (temp-pos);
            l = alfa[pos];
        }
        return l;
    }
	/* Stampo la frase risultante */
    public void stampa() {
    	System.out.println("Frase risultante: ");
        for(int i=0;i<getPhrase().length();++i) System.out.print(Risultato[i]);
    }
}

File: C++/vigenere.cpp
#include "vigenere.h"
#include <iostream>
/* Controllo se la chiave inserita rispetta i vincoli imposti (ovvero se è costituita da soli caratteri alfabetici) e in
 * caso affermativo trasformo ogni lettera da minuscolo a maiuscolo */
    Vigenere::Vigenere(string k,string p):Frase(p),Key(k){
        if(Key == "") Frase::warning("Non hai inserito la chiave!");
        else{
            int sizeK = Key.size();
            bool s = true;
            for(int i=0;i<sizeK && s;i++){
                if((Key.at(i)>='a' && Key.at(i)<='z') || (Key.at(i)>='A' && Key.at(i)<='Z')){
                    Key.at(i) = toupper(Key.at(i));
                }
                else{
                    Frase::warning("La chiave deve contenere solo caratteri alfabetici senza spazi!");
                    Key = "";
                    s = false;
                }
            }
        }
    }
/* Cifro il testo in chiaro */
    void Vigenere::codifica() {
        buildStringVig();
        string f = getPhrase();
        int sizeF = f.size();
        for(int i=0;i<sizeF;i++){
            if(S[i] != 0)
                Risultato[i] = getLettera(f.at(i),S[i]);
            else Risultato[i] = f.at(i);
        }
    }
/* Decifro il testo cifrato */
    void Vigenere::codificaRev() {
        buildStringVig();
        string f = getPhrase();
        int sizeF = f.size();
        for(int i=0;i<sizeF;i++){
            if(S[i] != 0)
                Risultato[i] = getLetteraRev(f.at(i),S[i]);
            else Risultato[i] = f.at(i);
        }
    }
/* Costruisco la stringa di cifratura e l'array che conterrà il numero di spostamenti da effettuare per ciascuna lettera
   del testo inserito */
    void Vigenere::buildStringVig(){
        string f = getPhrase();
        char* Lett = getAlfa();
        int sizeF = f.size();
        int sizeK = Key.size();
        for(int i=0,j=0;i<sizeF;i++){
            if(f.at(i)!=' '){
                if(j>=sizeK) j=0;
                CipherString[i] = Key.at(j);
                j++;
            }
            else CipherString[i] = ' ';
        }
        for(int i=0;i<sizeF;i++){
            if(CipherString[i]!=' '){
                for(int j=0;j<26;j++){
                    if(CipherString[i] == Lett[j]) S[i] = j;
                }
            }
            else S[i] = 0;
        }
    }
/* Trovo e ritorno la lettera che si trova 'pos' posizioni prima rispetto a 'c' */
    char Vigenere::getLetteraRev(char c,int pos) const{
        char l;
        char* alfa = getAlfa();
        int temp;
        for(int i=0;i<26;i++){
            if(alfa[i]==c){
                temp = i;
                break;
            }
        }
        if(temp >= pos)
            l = alfa[temp-pos];
        else{
            pos = 26 + (temp-pos);
            l = alfa[pos];
        }
        return l;
    }
/* Ritorno il testo risultante */
    char* Vigenere::getRisultato() const{
        char* p = new char[getPhrase().size()+1];
        strcpy(p,Risultato);
        return p;
    }
/* Ritorno la stringa di cifratura */
    char* Vigenere::getCipherString() const{
        char* p=new char[getPhrase().size()+1];
        strcpy(p,CipherString);
        return p;
    }
/* Ritorno la chiave */
    string Vigenere::getKey() const{ return Key; }
